### PR TITLE
build: sync XS with upstream

### LINKED ISSUE #3910: expose XS the->currentHeapCount to metering/delivery results
## What is the Problem Being Solved?

It'd be awfully helpful to have xsnap delivery results include `the->currentHeapCount`, as described in https://github.com/Agoric/agoric-sdk/issues/2631#issuecomment-922007282

This tracks the number of "slots" allocated by the XS heap. Each slot uses 32 bytes (on my 64-bit kernel), and most native JS objects need one or two of them. The number grows whenever an object is created (even temporarily), and only drops when GC happens. This is close enough to what we want a spacebank to measure that we should expose it so we can start paying attention to it.

It probably doesn't include "bulk" memory (used for strings, the backing store on an Array, the hash tables in Maps/Sets, also I think large `BigInt`s use a slab), so we'll need to find the right way to track that too.

`the->allocatedSpace` tracks a high-water mark of slots, never goes down, is denominated in bytes, and grows in much larger steps (maybe 128kiB? 1 MiB?). It tracks how much memory (for slots) XS has requested/received from the operating system, formerly via `malloc` and now through `mmap`. We currently report `the->allocatedSpace` in the metering results as `allocate`. We should keep reporting it, for now.

The task is to add `slots` to the metering results (which currently contain `compute` and `allocate`).

## Security Considerations

The number of slots allocated during execution is just as revealing as the number of computrons consumed, so the security profile doesn't change. Metering results are initially private to the kernel, but exposed (via Meters) to the controlling vat through the admin facet.

## Test Plan

`packages/xsnap` should acquire some more unit tests, next to the existing ones for `compute`, to ensure that the C code is including the metering results.

CC @dckc : if you feel you can take this on, go for it, else assign me and I'll see what I can figure out. We should check our work with the Moddable folks. Also, perhaps this ticket should be filed/transferred to the xsnap-pub repo, I wasn't sure.

### PR DESCRIPTION
**Pull request by dckc**
We don't have immediate plans to use petrify, resizable buffers, etc. but this is an important time to be in sync.

[xs-update.diff](https://gist.github.com/dckc/8ad0a7b473edef255952098d2bf7cfe1) is the output of `git diff 10cc52e130..7276ac7ba2d3` in the [xs](https://github.com/Moddable-OpenSource/moddable/tree/public/xs) subdirectory.

fixes #3910

Things todo before taking this out of draft mode:
 - [x] resolve two test failures
 - [x] wait for referenced xsnap-native commit to eventually appear

### CODE REVIEW

File: packages/xsnap/test/test-boot-lockdown.js [hunk 1]
```diff
@@ -186,9 +186,15 @@ test('console - objects should include detail', async t => {
...
   await worker.evaluate(`
-    const skipLineNumbers = s => !s.startsWith('Error: ');
-    send(printed.map(args => args.map(a => a.toString()).filter(skipLineNumbers)))
+  const skip = s =>
+    !s.startsWith('Error: ') && !s.startsWith('Removing intrinsics.');
+  send(
+    printed
+      .map(args => args.map(a => a.toString()).filter(skip))
+      .filter(a => a.length > 0),
+  );
```
Comment (erights) [inline packages/xsnap/test/test-boot-lockdown.js]: I did not know about this filter and I don't quite understand it, either before or afterwards. No change suggested. Given its purpose, don't block on my lack of understanding, but I'd appreciate discussing this sometime.

File: packages/xsnap/test/test-boot-lockdown.js [hunk 2]
```diff
@@ -200,7 +206,6 @@ test('console - objects should include detail', async t => {
           'assertion text',
           '{"prop1":["elem1a","elem1b"],"prop2":["elem2a","elem2b"]}',
         ],
-        [],
```
Comment (erights) [inline packages/xsnap/test/test-boot-lockdown.js:203]: What do the two deleted empty arrays correspond to?

**Review by erights — APPROVED** [review]
LGTM but would still be good to get @kriskowal 's and/or @warner 's reaction first

File: packages/xsnap/test/test-boot-lockdown.js (hunk 1 above)
Comment (dckc) [reply packages/xsnap/test/test-boot-lockdown.js]: This is a regression test for a situation where rich object structure showed up on the console as just `[object Object]`. It works by intercepting `print()`, but lockdown's warnings about removing unrecognized intrinsics also go via `print()`, and the recent XS update has some new stuff such as `String.prototype.at`. It looks like we fixed that particular case (https://github.com/endojs/endo/issues/854 ), but there are others, so I'm filtering them out for the purpose of this test.

File: packages/xsnap/test/test-boot-lockdown.js (hunk 2 above)
Comment (dckc) [reply packages/xsnap/test/test-boot-lockdown.js:203]: `print()` calls where all the arguments got filtered. The lockdown warnings introduced several more, so I filtered them all out.

Comment (dckc) [reply packages/xsnap/test/test-boot-lockdown.js:203]: p.s. I'm embarrased to explain this code. It clearly merits refactoring.

**Review by kriskowal — APPROVED** [review]

File: packages/xsnap/test/test-boot-lockdown.js (hunk 1 above)
Comment (dckc) [reply packages/xsnap/test/test-boot-lockdown.js]: I refactored this bit to be more clear.

**Pull request merged**